### ISSUE TITLE
Cannot export or import all frames or shapes in shapes.vga for shapes numbered 149 and below using ExultStudio.

### ISSUE THREAD

**Issue by GrantWChapman** [exult/exult#577]
Is it possible to add the ability in ExultStudio to export and import all frames or shp files in shapes numbered under 150?

**Comment by KnightCaptainU7** [comment]
Can confirm this bug in Studio 1.10.1. Right-clicking the shapes from 0-149 does not provide "all frames" options.
![image](https://github.com/user-attachments/assets/8feb2fbf-cd9b-4d22-94e9-b92400e628b5)

Shapes starting at 150 have the "all frames" options.
![Screenshot 2024-11-29 184522](https://github.com/user-attachments/assets/5b8fc44b-ee6b-4169-86f5-367fce9f70e7)

Both BG and SI are impacted. Possibly hardcoded in Studio?

**Comment by KnightCaptainU7** [comment]
The "Edit tiled..." option should be "Export tiles" but it is not clear to me why that should be a different dialog box that only does 4 at a time?

**Comment by DominusExult** [comment]
I remember that I brought this up years ago and I think @wjp said that these shapes are special 🤷‍♂️
But I don't remember what exactly he said and where. Couldn't find it on SF bug tracker or the old forum.

**Comment by Dragon-Baroque** [comment]
> these shapes are special.

Indeed, they are the Ground _Tiles_. All are 8x8, no offset, no height.

Although they are multi-frames, like the regular shapes, the _Edit tiled..._ takes over the _Export All Frames_ by feeding Gimp directly a band of height 8 where the various frames are drawn beside each other ( by row ), or a band of width 8 where the various frames are drawn below each other ( by column ).

Admittedly, instead of 4 frames by default, _Edit tiled..._ might put the number of frames of the Tile Shape ...

**Comment by KnightCaptainU7** [comment]
> Although they are multi-frames, like the regular shapes, the _Edit tiled..._ takes over the _Export All Frames_ by feeding Gimp directly a band of height 8 where the various frames are drawn beside each other ( by row ), or a band of width 8 where the various frames are drawn below each other ( by column ).

It's a cool feature, so it's great that is there. It shouldn't prevent the use of _Import/Export All Frames_ because that limits us just to GIMP.

Use cases for showing _Import/Export All Frames_ include:
Wanting to use some other graphics editor, such as pixel-focused editors.
Wanting to work with the frames as individual files, as I do for recolors via ImageMagick.
Wanting to re-import them in a custom game, such as a BG+SI hybrid.
Easier documentation, it's better to have consistent features and then include bonus features.

**Comment by Dragon-Baroque** [comment]
> It shouldn't prevent the use of Import/Export All Frames because that limits us just to GIMP.

GIMP is the default value of the configuration property `config/estudio/image_editor`, which you can alter, and which also is reached within Studio by `File -> Preferences... -> Image Editor Command`.

> Use cases for showing `Import/Export All Frames` include:
Wanting to use some other graphics editor, such as pixel-focused editors.
Wanting to work with the frames as individual files, as I do for recolors via ImageMagick.
Wanting to re-import them in a custom game, such as a BG+SI hybrid.
Easier documentation, it's better to have consistent features and then include bonus features.

Good reasons, I yield. But this is an area where I have little practice. If you are willing to help me, and you are able to build Exult from sources, here is a proposal :
- Grab Exult from master,
- Alter the if statement in `mapedit/shapelst.cc` line 2241..2242 from
` if (!IS_FLAT(info[selected].shapenum) || file_info != studio->get_vgafile()) { ` to 
` if (file_info != studio->get_vgafile()) { `. __Explanation:__ A Tile Shape is called a _flat_ in that code. This change enables the `Export All Frames` and `Import All Frames` on `Tile` Shapes as you wish.
- Play with the `Export All Frames` and `Import All Frames` onto `Tile` Shapes and check that it performs as expected.
- Now for the tricky part. `Import All Frames` probably does not expect to run onto a Tile Shape, and probably does not check that all Frames are exactly 8x8. Could you devise a test whereby you try to `Import All Frames` from a Shape with a larger size ? If `Import All Frames` accepts that - and thus damages the `shapes.vga`, then I have to alter the `Import All Frames` code to add a size check.

**Comment by Dragon-Baroque** [comment]
> Is it possible to add the ability in ExultStudio to export and import all frames or shp files in shapes numbered under 150?

@GrantWChapman, @KnightCaptainU7, this is in Studio from this morning by PR #758.
Could you test whether it is satisfactory ? Can we close the issue ?

**Comment by DominusExult** [comment]
I forgot to mention a vital point. Yes, it shows an empty image but you can see that there is something else outside the canvas (yellow square, to the top left of the empty canvas). So in Gimp's menu -> image -> fit canvas to layer and the images are shown.

**Comment by Dragon-Baroque** [comment]
This opens the ability to satisfy the `Export / Import Shape` requirement on Tiles. 
It means that the comment on `Shape::load` : `Load all frames for a single shape.  (Assumes RLE-type shape.)` is wrong, `Shape::load` can load Tile shapes too.

Then it becomes only a matter of unlocking the `Export Shape` and `Import Shape` in the contextual menu of a Shape, and to alter the checks on the incoming Shape files :
- Accept n * 64 sized Shape files in Return from Edit ,
- Depending on the Flatness of the current Shape, accept either n * 64 sized Shape files or RLE Shape files.

Here is a patch as a first attempt to that.

Two problems :
- It does not allow to Import an even right sized ( 8 x 8 ) RLE exported Shape file into a Tile Shape, nor to Import a Tile Shape export into a regular Shape, I do not know whether that should be considered important,
- Both the Edit and the Import Shape appear to work against Tiles, but they behave strangely in All Frames mode. It looks that the Bounding Box array of the Edited / Imported Shape is not properly refreshed. Does seem to occur with regular Shapes as well.

[ExportImportTileShape.zip](https://github.com/user-attachments/files/24371105/ExportImportTileShape.zip)

**Comment by DominusExult** [comment]
With this patch I was able to export/import SHP on these shapes (and using "Edit" works now as well that edited changes are read in).

I have to confess I don't really understand what you mean with the two problems.

**Comment by DominusExult** [comment]
Now I understand :)
1. IMO leave it as it is.

2. I see. Probably something we just have to live with.
